### ISSUE TITLE
Unexpected scalar x array behavior

### ISSUE THREAD

**Issue by mishpat** [numba/numba#2783]
In the code below I get f1 returning an empty array on Numba 0.36.1, Python 3.6.1, Windows 10; of course it would seem to me they are all equivalent.
[]

[[ 0.13447071  0.        ]
 [ 0.          0.13447071]]

[[ 0.13447071  0.        ]
 [ 0.          0.13447071]]

```
@jit(nopython=True)
def f1(arr):
    b = 1.0
    out = (b * arr)/np.sum(np.exp(b * arr))
    return out

@jit(nopython=True)
def f2(arr):
    b = 1.0
    k = b * arr
    out = k/np.sum(np.exp(b * arr))
    return out

@jit(nopython=True)
def f3(arr):
    out = arr / np.sum(np.exp(arr))
    return out
x = np.eye(2)
print(f1(x))
print(f2(x))
print(f3(x))
```

I can't read the IR very well but I don't see any obvious problem.

**Comment by stuartarchibald** [comment]
Thanks for the report, I can reproduce. I think it's something in the array expr rewrite pass that is causing it. The problem is array expr invariant:
```python
from numba import njit
import numpy as np

def f(arr):
    return (1 + arr) + np.sum(1 + (1  + arr))

x = np.ones((2,3))
print(f(x))
print(njit(no_rewrites=True)(f)(x))
print(njit(no_rewrites=False)(f)(x))
```
gives:
```python
[[ 20.  20.  20.]
 [ 20.  20.  20.]]
[[ 20.  20.  20.]
 [ 20.  20.  20.]]
[]
```

**Comment by stuartarchibald** [comment]
This is fixed as of 0.45. Closing.
